lib/Reader/readerir.cpp

```diff
@@ -7666,97 +7706,186 @@ Type *GenIR::Vector4Ty;
+    return (IRNode *)LLVMBuilder->CreateMul(Vector1, Vector2);
+  } else {
+    return 0;
+  }
 }
 
-IRNode *GenIR::vectorDiv(IRNode *Vector1, IRNode *Vector2) {
-  assert(((Value *)Vector1)
-             ->getType()
-             ->getVectorElementType()
-             ->isFloatingPointTy());
-  assert(((Value *)Vector2)->getType() == ((Value *)Vector1)->getType());
-  return (IRNode *)LLVMBuilder->CreateFDiv(Vector1, Vector2);
+IRNode *GenIR::vectorDiv(IRNode *Vector1, IRNode *Vector2,
+                         CORINFO_CLASS_HANDLE Class) {
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    return (IRNode *)LLVMBuilder->CreateFDiv(Vector1, Vector2);
+  } else if (Vector1->getType()->getVectorElementType()->isIntegerTy()) {
+    int Length = 0;
+    bool IsGeneric = false;
+    bool IsSigned = false;
+    getBaseTypeAndSizeOfSIMDType(Class, Length, IsGeneric, IsSigned);
+    if (IsSigned) {
+      return (IRNode *)LLVMBuilder->CreateSDiv(Vector1, Vector2);
+    } else {
+      return (IRNode *)LLVMBuilder->CreateUDiv(Vector1, Vector2);
+    }
+  } else {
+    return 0;
+  }
 }
 
 IRNode *GenIR::vectorEqual(IRNode *Vector1, IRNode *Vector2) {
-  assert(((Value *)Vector1)
-             ->getType()
-             ->getVectorElementType()
-             ->isFloatingPointTy());
-  assert(((Value *)Vector2)->getType() == ((Value *)Vector1)->getType());
-  return (IRNode *)LLVMBuilder->CreateFCmpOEQ(Vector1, Vector2);
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    return (IRNode *)LLVMBuilder->CreateFCmpOEQ(Vector1, Vector2);
+  }
+  if (Vector1->getType()->getVectorElementType()->isIntegerTy()) {
+    return (IRNode *)LLVMBuilder->CreateICmpEQ(Vector1, Vector2);
+  }
+  return 0;
 }
 
 IRNode *GenIR::vectorNotEqual(IRNode *Vector1, IRNode *Vector2) {
-  assert(((Value *)Vector1)
-             ->getType()
-             ->getVectorElementType()
-             ->isFloatingPointTy());
-  assert(((Value *)Vector2)->getType() == ((Value *)Vector1)->getType());
-  return (IRNode *)LLVMBuilder->CreateFCmpONE(Vector1, Vector2);
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    return (IRNode *)LLVMBuilder->CreateFCmpONE(Vector1, Vector2);
+  }
+  if (Vector1->getType()->getVectorElementType()->isIntegerTy()) {
+    return (IRNode *)LLVMBuilder->CreateICmpNE(Vector1, Vector2);
+  }
+  return 0;
+}
+
+IRNode *GenIR::vectorMax(IRNode *Vector1, IRNode *Vector2) {
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    std::vector<Type *> Types;
+    Types.push_back(Vector1->getType());
+    llvm::Function *Func = 0;
+    Func = Intrinsic::getDeclaration(JitContext->CurrentModule,
```

Comment:
I'm not sure that `llvm.maxnum` is the right thing to use here: the [documentation](http://llvm.org/docs/LangRef.html#llvm-maxnum-intrinsic) does not indicate that it supports vector arguments. I think that the right thing to do is to generate IR that looks something like this (and similar for integer vectors):

```
%vcmpres = fcmp ogt <n x f> %vector1, %vector2
%vresult = select <n x i1> %vcmpres, <n x f> %vector1, <n x f> %vector2
```